VSR: Request dirty prepare

sentientwaffle opened the pull request:
Improve availability of `on_request_prepare` by always calling `Journal.read_prepare_with_op_and_checksum`, not `Journal.read_prepare` — even when the header is in memory. This prevents a deadlock when:
  1. Replica 2 has op=26, but it is faulty.
  2. Replica 1 does not have op=26.
  3. Replica 0 has op=26, but after crashing and recovering, op=26 was recovered as `fix` (i.e. dirty but not faulty).

One thing to note is that `request_prepare` will now keep trying to read the prepare from the WAL _even if the slot is marked faulty_. It did that before too, but only when there was no in-memory header.

See https://github.com/coilhq/tigerbeetle/issues/109#issuecomment-1201798385 for more detail.

---

Other changes:

- WAL: Move the `message.size == @sizeOf(Header)` optimization from `read_prepare` to `read_prepare_with_op_and_checksum` so that `on_request_prepare` can take advantage of it.
- Simulator: Log when storage faults are enabled/disabled per replica.
- Tinker with a few log messages/levels.
- Add an unrelated assertion to `Journal.read_prepare_with_op_and_checksum`.
- Remove dead code in `count_quorum` (unused because `start_view_change` and `nack_prepare` use `QuorumCounter`s instead of `QuorumMessages`. Move their assertions to `count_message_and_receive_quorum_exactly_once`.
- Remove invalid assertion from `repair_header`. It was only not failing because the WAL doesn't wrap yet.

sentientwaffle committed:
VSR: Fix request_prepare deadlock

Improve availability of `on_request_prepare` by always calling `Journal.read_prepare_with_op_and_checksum`, not `Journal.read_prepare` — even when the header is in memory. This prevents a deadlock when:
  1. Replica 2 has op=26, but it is faulty.
  2. Replica 1 does not have op=26.
  3. Replica 0 has op=26, but after crashing and recovering, op=26 was recovered as `fix` (i.e. dirty but not faulty).

One thing to note is that `request_prepare` will now keep trying to read the prepare from the WAL _even if the slot is marked faulty_.

See https://github.com/coilhq/tigerbeetle/issues/109 for more detail.

Other changes:

- WAL: Move the `message.size == @sizeOf(Header)` optimization from `read_prepare` to `read_prepare_with_op_and_checksum` so that `on_request_prepare` can take advantage of it.
- Simulator: Log when storage faults are enabled/disabled per replica.
- Tinker with a few log messages/levels.
- Add an unrelated assertion to `Journal.read_prepare_with_op_and_checksum`.

sentientwaffle committed:
VSR: Remove dead code, invalid assertion

- Remove dead code in `count_quorum` (unused because `start_view_change` and `nack_prepare` use `QuorumCounter`s instead of `QuorumMessages`. Move their assertions to `count_message_and_receive_quorum_exactly_once`.
- Remove invalid assertion from `repair_header`. It was only not failing because the WAL doesn't wrap yet.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2004,13 +1994,16 @@ pub fn Replica(
                     assert(self.status == .view_change);
                     assert(self.leader_index(self.view) == self.replica);
+                    assert(self.replica != message.header.replica);
+                    assert(self.nack_prepare_op.? == message.header.op);
```
Let's put `message` on the left hand side of these assertions, since the subject (the thing that we're checking) is the message, and since there's also more shared prefix in `message.header` for things to line up vertically.

jorangreef reviewed (approved):
Awesome find and fix @sentientwaffle! Do you think we could get rid of `read_prepare()` entirely? And then be more flexible in reading from disk also when committing, with an assertion rather to ensure that the header wasn't changed concurrently during commit?

sentientwaffle committed:
WAL: read_prepare: Don't require entry to be clean

`Journal.read_prepare` is called by `Replica` when repairing the pipeline of a new leader, or reading an op to commit. It previously required that the entry be clean (`!dirty` and `!faulty`). This commit replaces that requirement with a looser one: `prepare_inhabited[slot] && prepare_checksums[slot] == requested_checksum`.

This allows a replica to:
- commit ops that were recovered with decision=`fix` (`dirty && !faulty`) since they have an intact prepare
- commit ops that were recovered with decision=`vsr` (`dirty && faulty`) when the prepare is intact (decisions `@E`, `@G`, `@L`, `@M`).

This is safe — the replica wouldn't be committing without a valid hash chain.

Other changes:
- Also, reorder `Replica` asserts to emphasize the subject.

See discussion: https://github.com/coilhq/tigerbeetle/pull/126.

sentientwaffle committed:
VSR: Make pre-commit checks stricter

In `commit_ops_commit`, the checks for `prepare.?.header.op != op` and `prepare.?.header.checksum != …`) are dead code, and can actually be asserts. (They are combined into one check with `assert(journal.has(prepare.?.header))`).

This works because:

- `read_prepare_with_op_and_checksum_callback` verifies that it is returning the requested op/checksum.
- And the (soon-to-be-committed) header in `journal.headers` should never be rewritten before its commit finishes.

See discussion: https://github.com/coilhq/tigerbeetle/pull/126.

sentientwaffle commented:
> Do you think we could get rid of `read_prepare()` entirely?

It is still useful to keep `read_prepare()` around because it enforces the matching `journal.headers[slot]`.

But it can be simplified: Drop the `!dirty` and `!faulty` requirements from `read_prepare()` in favor of just checking `prepare_inhabited[slot]` and `prepare_checksums[slot] == expected_checksum`.

> with an assertion rather to ensure that the header wasn't changed concurrently during commit?

:+1:

jorangreef reviewed (approved):
Nice work on spotting the dead code and asserting on that instead.
